### PR TITLE
Prefer salt.utils.stringutils.expr_match() over re.search() in state.event.

### PR DESCRIPTION
**Pull request by ghost**
### What does this PR do?

Replaces `fnmatch.match()` in `state.event` (`modules/state.py`) with `salt.utils.stringutils.expr_match()`.

A few merits of this approach have been discussed in #46554 and #46656: while @isbm initially supported a disruptive approach, @terminalmage pointed out that breaking the expectations of users is not desired if it can be avoided. So, the enlightened public seems to gravitate towards a non-disruptive approach, which is what this PR now tries to realise.

### What issues does this PR fix or reference?

Sometimes it gets rather tedious to debug event-related problems (ie. broken orchestrator states and so on) on a master which is really busy (200+ events per second) without a decent matching capability to filter events. Giving a possibility to use regular expressions to filter events displayed makes the life of an administrator a tad easier.

### Previous Behavior

The function `state.event()` used `fnmatch.match()` to filter events.

### New Behavior

The function `state.event()` uses `salt.utils.stringutils.expr_match()` to filter events.

### Tests written?

No (`state.event()` neither returns a value nor does it have any side effects).

### Commits signed with GPG?

No
See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/modules/state.py
Comment (terminalmage) [inline salt/modules/state.py]: Can you explain your reasoning for removing the `str()` here? It was done for a very good reason. With `unicode_literals` imported, and `json.dumps()` producing a `str` type, the format string has to also be a `str` type to avoid a decode error when the dumped data includes unicode characters:

```python
>>> from __future__ import unicode_literals
>>> import salt.utils.json
>>> data = {'спам': 'яйца'}
>>> data
{u'\u0441\u043f\u0430\u043c': u'\u044f\u0439\u0446\u0430'}
>>> '{0}'.format(salt.utils.json.dumps(data))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xd1 in position 2: ordinal not in range(128)
>>> str('{0}').format(salt.utils.json.dumps(data))
'{"\xd1\x81\xd0\xbf\xd0\xb0\xd0\xbc": "\xd1\x8f\xd0\xb9\xd1\x86\xd0\xb0"}'
>>>
```

This should be reverted back to the way it was.

File: salt/modules/state.py
Comment (terminalmage) [inline salt/modules/state.py]: This is a good change, it prevents a bunch of confusing debugging messages that I've seen from time to time as `state.event` is used.

**Review by terminalmage — CHANGES_REQUESTED** [review]

File: salt/modules/state.py
Comment (samilaine) [reply salt/modules/state.py]: I'm ashamed to admit that the only reason for removing `str()` because @isbm asked rather suggestively why it's there. Ie. the reason is that more attention got paid for who is saying that what was being said.

Comment (isbm) [reply salt/modules/state.py]: @samilaine @terminalmage *\*ahemmm\**, I didn't really meant "please remove it and never use it again". I just thought that `salt.utils.stringutils.to_str` is more correct to use here instead of just `str`.

Comment (samilaine) [reply salt/modules/state.py]: Communication between humans usually fails and when it doesn't, it's an accident. By no means I tried to lay blame on anyone else, as it was yours truly who failed to comprehend that "please remove it and never use it again" wasn't what was meant.

Comment (isbm) [reply salt/modules/state.py]: :wink: no problem, take it easy. The `str` is OK, but my perfectionist guts thinks that using standard utility for this would be more "kosher" here, since it does things according to the current Python version. I would leave this to @terminalmage — he is the Unicode boss here.

Comment (terminalmage) [reply salt/modules/state.py]: `to_str()` is only necessary when it's possible that the string may contain unicode, as it will perform the necessary `.encode()` on PY2 and do nothing on PY3.

Given that we know that the format string is ascii-only, `str()` is fine here.

Comment (terminalmage) [reply salt/modules/state.py]: @samilaine If you can get that line reverted back to what it was before, then this should be good to merge.

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by samilaine** [comment]
Failing to see how to paddle out of the swampy mess created by an untimely rebase, a new PR 46861 has been made to replace this one.

**Pull request closed without merging**
